File: 2130.maximum-twin-sum-of-a-linked-list.py
#
# @lc app=leetcode id=2130 lang=python
#
# [2130] Maximum Twin Sum of a Linked List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def pairSum(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: int
        """
        l = []
        left = 0
        right = -1
        while head is not None:
            l.append(head.val)
            head = head.next
            right += 1
        ts = -float('inf')
        while left < right:
            ts = max(ts, l[left] + l[right])
            left += 1
            right -= 1
        return ts
        
# @lc code=end

File: 2130.maximum-twin-sum-of-a-linked-list.java
/*
 * @lc app=leetcode id=2130 lang=java
 *
 * [2130] Maximum Twin Sum of a Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int pairSum(ListNode head) {
        ListNode slow=head;
        ListNode fast=head;
        while(fast!=null && fast.next!=null){
            slow=slow.next;
            fast=fast.next.next;
        }
        ListNode prev=null;
        ListNode curr=slow;
        while(curr!=null){
            ListNode newNode=curr.next;
            curr.next=prev;
            prev=curr;
            curr=newNode;
        }
        int maxSum=0;
        ListNode first=head;
        ListNode second=prev;
        while(second!=null){
            int sum=first.val+second.val;
            maxSum=Math.max(maxSum,sum);
            first=first.next;
            second=second.next;
        }
        return maxSum;
    }
}
// @lc code=end

